servers/rest/rest.go

```diff
@@ -2,54 +2,107 @@
 package rest
 
 import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
 	"net"
 	"net/http"
+	"time"
 
 	"github.com/gluster/glusterd2/middleware"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/gorilla/mux"
 	"github.com/justinas/alice"
 	"github.com/soheilhy/cmux"
+	config "github.com/spf13/viper"
 )
 
 // GDRest is the GlusterD Rest server
 type GDRest struct {
 	Routes   *mux.Router
 	listener net.Listener
+	server   *http.Server
+	stopCh   chan struct{}
 }
 
-// New returns a GDRest object which can listen on the configured address
-func New(l net.Listener) *GDRest {
-	rest := &GDRest{
-		mux.NewRouter(),
-		l,
+func tlsListener(l net.Listener) (net.Listener, error) {
+	certificate, err := tls.LoadX509KeyPair(
+		config.GetString("client-cert-file"),
+		config.GetString("client-key-file"))
+	if err != nil {
+		return nil, err
 	}
 
-	rest.registerRoutes()
+	config := &tls.Config{
+		Certificates: []tls.Certificate{certificate},
+		Rand:         rand.Reader,
+	}
 
-	return rest
+	return tls.NewListener(l, config), nil
 }
 
 // NewMuxed returns a GDRest object which listens on a CMux multiplexed connection
 func NewMuxed(m cmux.CMux) *GDRest {
-	return New(m.Match(cmux.HTTP1Fast()))
+
+	rest := &GDRest{
+		Routes: mux.NewRouter(),
+		server: &http.Server{},
+		stopCh: make(chan struct{}),
+	}
+
+	enableSSL := true
+	if l, err := tlsListener(m.Match(cmux.TLS())); err != nil {
+		log.WithError(err).Warn("Failed to create SSL/TLS listener. Falling back to HTTP")
+		enableSSL = false // fallback to not using TLS
```

Comment:
Don't fallback to an not using TLS. If a user has requested TLS and you cannot start a TLS listener, fail.